Update Artifactory package pull-through doc (#2544)
chainguard-dev/edu: content/chainguard/chainguard-images/chainguard-registry/pull-through-guides/artifactory/artifactory-packages-pull-through/index.md

Set your Chainguard organization identifier as an environment variable. Replace the `example.org` placeholder with your organization's name as it appears in the Chainguard Console:

```shell
export CHAINGUARD_ORG=example.org
```

Next, generate a pull token:

```shell
chainctl auth pull-token --parent=${CHAINGUARD_ORG} -o env
```

The command outputs the credentials as `export` commands you can run to set them as environment variables:

```output
export CHAINGUARD_IDENTITY_ID=<id>
export CHAINGUARD_TOKEN=<token>
```

Run the first of these commands to create an environment variable for the identity:

```shell
export CHAINGUARD_IDENTITY_ID=<id>
```

Be sure to note both the identity and token values to use when you create a remote Artifactory repository in the next section.

> With the environment variables in place, create a role binding to grant the pull token identity the `apk.pull` role:

Author: Good catch, I think just "bind to the pull token identity" makes the most sense here

### Generating a token for the remote repository

Before testing whether you're able to pull packages through the remote Artifactory repository, you must retrieve a token generated for it by Artifactory.

1. On the Repositories page, find your repository.
2. Click the ellipsis (**⋯**) all the way to the right of the repository's row.
3. Select **Set me up**. 
4. Click the **Generate Token & Create Instructions** button.
5. Copy or take note of the token generated for your repository.

With this token in hand, run the following `export` command to create an environment variable, replacing `my-token` with the token you noted down. You will use this environment variable in the following section as you test out the remote repository:

```shell
export CG_PRIVATE_TOKEN=my-token
```

Additionally, create two more environment variables to hold your Artifactory user profile and hostname:

```shell
export ARTIFACTORY_USER_PROFILE=my-user-profile
export ARTIFACTORY_HOSTNAME=my-artifactory-hostname
```

If you aren't sure of these values, you can find them in the command from the **Set Up An Alpine Client** window where you found the token:

```shell

> sudo sh -c "echo 'https://linky:<TOKEN>@example-hostname.jfrog.io/artifactory/cg-private/<BRANCH>/<REPOSITORY>'" >> /etc/apk/repositories

Author: Based on my read of the docs "server name" is the term we're looking for here. Good catch!

After creating both repositories, generate and retrieve a token for each one as you did for the `cg-private` remote repository:

1. On the Repositories page, find both remote repositories you just created.
2. Click the ellipsis (**⋯**) all the way to the right of their respective rows.
3. Select **Set me up**. 
4. Click the **Generate Token & Create Instructions** button.
5. Copy or take note of the tokens generated for your repositories.

Next, copy the resulting tokens to create a pair of environment variables. For the `cg-chainguard` repository's token:

```shell
export CG_TOKEN=<cg-chainguard-token>
```

Then for the `cg-extras` repo's token:

```shell
export EXTRAS_TOKEN=<cg-extras-token>
```

Following that, you can use these environment variables, along with a few created earlier in the guide, to create a Dockerfile as you previously did for the `cg-private` repository:

```shell
cat > Dockerfile.repos <<EOF
FROM cgr.dev/chainguard/python:latest-dev
USER root
RUN mv /etc/apk/repositories /etc/apk/repositories.disabled
RUN echo 'https://$ARTIFACTORY_USER_PROFILE:$CG_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-chainguard/' >> /etc/apk/repositories

> RUN echo 'https://$ARTIFACTORY_USER_PROFILE:$EXTRAS_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-extras/' >> /etc/apk/repositories

Reviewer: I think it would actually be best to use three commands. First one to add public and overwrite the file with > 

and then extras with >> so its appended
and then maybe even again >> for private repo

weight: 010
---

This tutorial outlines how to set up remote Alpine package repositories with [JFrog Artifactory](https://jfrog.com/artifactory/) that can be used as pull-through caches for Chainguard package repositories. Specifically, it will walk you through how to set up remote Artifactory repositories to serve as pull-through caches for a Chainguard private APK repository as well as Chainguard's public package repositories.

This guide will then outline how you can configure an image to pull APK packages from these remote repositories using tokens generated by Artifactory. It will also go over how you can use a virtual repository from Artifactory as a pull-through cache for Chainguard packages.

## Prerequisites

In order to complete this tutorial, you will need the following:

> * Administrative privileges over an Artifactory instance. If you're interested in testing out this configuration, you can set up a trial instance on the [JFrog Artifactory landing page](https://jfrog.com/artifactory/).

Author: fair!

> > **Note**: As you follow this guide, you may run into issues if your Artifactory username is an email address; specifically, the `@` sign can lead to errors. For simplicity's sake, we recommend that you use a user profile with a name that only contains letters and numbers. If you must use an email address as your Artifactory username, though, this guide will include instructions for working around the issue.

Author: I think this was more from the perspective of helping out my fellow Chainguardians with reviewing, good point though. Will remove the note.

## Creating a Chainguard Pull Token for Remote Repository Authentication

When configuring an Artifactory remote repository to function as a pull-through cache for packages from a Chainguard [Chainguard Private APK Repository](/chainguard/chainguard-images/features/private-apk-repos/), the remote repository will need to authenticate to Chainguard in order to reach the private repository. This section outlines the steps necessary to create a Chainguard pull token and configure the required permissions to access your Chainguard organization's private APK repository.

First, set your Chainguard organization identifier as an environment variable. Be sure to replace the `example.org` placeholder with your organization's name as it appears in the Chainguard Console:

```shell
export CHAINGUARD_ORG=example.org
```

Next, generate a pull token:

```shell
chainctl auth pull-token --parent=${CHAINGUARD_ORG} -o env
```

This command creates a new pull token identity within your specified organization and outputs the credentials as `export` commands you can run to set them as environment variables:

```output
export CHAINGUARD_IDENTITY_ID=<id>
export CHAINGUARD_TOKEN=<token>
```

> Run the first of these commands to create an environment variable for the identity ID:

Author: ah yeah, that is

```shell
export CHAINGUARD_IDENTITY_ID=<id>
```

You don't need to create the `CHAINGUARD_TOKEN` variable for the authentication token, but you should note the token value, as you'll need it when you create a remote Artifactory repository in the next section.

With the environment variables in place, create a role binding to grant the pull token identity the `apk.pull` role:

```shell
  chainctl iam role-bindings create \
    --parent=${CHAINGUARD_ORG} \
    --identity=${CHAINGUARD_IDENTITY_ID} \
    --role=apk.pull
```

> This command binds the `apk.pull` role to your newly created identity, giving it the permissions needed to download packages from your organization's private APK repository.

Author: I'm going to keep your language but leave it as two sentences, I try to avoid preceding a heading with a code block.

> This command binds the `apk.pull` role to your newly created identity, giving it the permissions needed to download packages from your organization's private APK repository.

Reviewer: Why is that?

* **Repository Key** — This is a name used to identify your remote repository. You can choose whatever name you like here, but this guide's examples will use the name `cg-private`.
* **URL** — This must be set to `https://apk.cgr.dev/${CHAINGUARD_ORG}`, but with your organization's actual name in place of `${CHAINGUARD_ORG}`. For example, if your organization is named `example.org` you would enter `https://apk.cgr.dev/example.org` here.
* **User Name** — This is used by Artifactory to authenticate to Chainguard and access your private APK repository. Here, add the pull token identity (the value you set to the `CHAINGUARD_IDENTITY_ID` variable earlier).
* **Password / Access Token** — This, along with the User Name, authenticates the Artifactory repository to access the remote private APK repository. Here, enter the token value you noted down after it was generated by the previous `chainctl auth pull-token` (`export CHAINGUARD_TOKEN=...`). 

Keep all the remaining fields set to their default values and click the **Create Remote Repository** button. This will create the remote repository and return you to the **Repositories** page. 

> ### Generating a token for the remote repository

Author: H2s are Title case, H3s are sentence case per the style guide: https://github.com/chainguard-dev/edu/wiki/Technical-Writing-Guidelines#headings

> ### Generating a token for the remote repository

Reviewer: ugh .. I wasnt aware of that .. thats terrible

> Before testing whether you're able to pull packages through the remote Artifactory repository you just created, you will need to retrieve a token generated for it by Artifactory.

Reviewer: ```suggestion
Before testing whether you're able to pull packages through the remote Artifactory repository, you must retrieve a token generated for it by Artifactory.
```

On the Repositories page, find your repository, click the ellipsis (**⋯**) all the way to the right of the repository's row, and then select **Set me up**. 

> This will open a window from the right side of the page. Click the **Generate Token & Create Instructions** button, which will generate two code blocks whose contents you can copy. The first code block will contain the token you can use to access the remote Artifact repository. The second block will contain an `echo` command you can use to add a URL to an `/etc/apk/repositories` file. 

Reviewer: Again .. keep it simpler and less wordy and just make a list of small steps.

> Copy or take note of the token generated for your repository. Then, run the following `export` commands to create an environment variable, replacing `my-token` with the token you noted down. You will use this environment variable in the next sections as you test out the remote repository:

Reviewer: same here

If you aren't sure of these values, you can find them in the command from the **Set Up An Alpine Client** window where you found the token:

```shell

> sudo sh -c "echo 'https://linky:<TOKEN>@example-hostname.jfrog.io/artifactory/cg-private/<BRANCH>/<REPOSITORY>'" >> /etc/apk/repositories

Reviewer: what is the branch and repository stuff about

Author: That's just what appears on Artifactory :shrug:

```

In this example, the Artifactory user profile is `linky` and the hostname is `example-hostname`.

> > **Note**: If your Artifactory user profile is an email address, you will encounter an error unless you percent-encode the `@` sign, as in `export ARTIFACTORY_USER_PROFILE=linky%40example.com`. Here, the Artifactory user profile is `linky@example.com`, but it must be entered into the environment variable as `linky%40example.com`. 

Reviewer: ```suggestion
> **Note**: If your Artifactory user profile is an email address, you must percent-encode the `@` sign, as in `export ARTIFACTORY_USER_PROFILE=linky%40example.com`. Here, the Artifactory user profile is `linky@example.com`, but it must be entered into the environment variable as `linky%40example.com`. 
```

Still a bit wordy but..

## Testing Pull-Through from Private APK Repository

> In order to test that you're able to pull packages from the remote Artifactory repository you just created, this section outlines how to build a container image (using a Chainguard image as a base) and configure it to pull packages from these repositories.

Reviewer: ```suggestion
This section outlines how to build a container image using a Chainguard image as a base and configure it to pull packages from the private APK repositories via Artifactory.
```

> ### Testing package pull-through interactively

Reviewer: Title case?

Reviewer: Also .. I think this whole section is redundant .. the next section where the apk install is in the dockerfile is good enough and you could add a verbose flag or so if you really want.. this should not be a docker tutorial

docker run -it --entrypoint /bin/sh ar-interactive
```

Now that the container is running, you can check whether it was configured correctly and you can pull packages from the remote Artifactory repositories.

First check that the `/etc/apk/repositories` file contains the correct URL:

```container
cat /etc/apk/repositories
```
```output
https://$ARTIFACTORY_USER_PROFILE:$CG_PRIVATE_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-private/
```

Then test that you can install a package from these repositories with `apk`. This example installs the `npm` package:

```container

> apk add npm

Reviewer: Not sure npm is a good example since its part of the JS libraries..

Author: getting rid of this section, so this is moot

```

> Be aware that you will only be able to add packages that are available from your organization's private APK repository. Refer to our [private APK repository documentation](/chainguard/chainguard-images/features/private-apk-repos/#about-private-apk-repositories) for more information.

Reviewer: ```suggestion
You can only add packages that are available from your organization's private APK repository. Refer to our [private APK repository documentation](/chainguard/chainguard-images/features/private-apk-repos/#about-private-apk-repositories) for more information.
```

Reviewer: Maybe even worth clarifying earlier and that this only applies to packages that are in your private repo

> This command's output shows that `npm` was indeed correctly installed from the remote Artifactory repository, which is itself pulling from the private APK repository:

Reviewer: ```suggestion
The output shows that `npm` was correctly installed from the remote Artifactory repository:
```

```output
fetch https://<ARTIFACTORY-USER-PROFILE>:*@<ARTIFACTORY-HOSTNAME>.jfrog.io/artifactory/cg-private/x86_64/APKINDEX.tar.gz
(1/1) Installing npm (11.4.2-r1)
Executing busybox-1.37.0-r49.trigger
OK: 679 MiB in 76 packages
```

You can confirm this from the Artifactory dashboard with the [**Artifact Repository Browser**](https://jfrog.com/help/r/jfrog-artifactory-documentation/browsing-artifacts). Navigate to the **Platform** tab, then find **Artifactory** in the left-hand navigation menu and select **Artifacts**. From there, expand the menu option for the `cg-private` Artifactory repository. There, you will find the `npm` package listed, along with any other packages you've pulled through the remote repository.

> ### Testing pull-through within a build

Reviewer: title case?

Reviewer: and maybe also simplify if you remove the earlier section

FROM cgr.dev/chainguard/python:latest-dev
USER root
RUN mv /etc/apk/repositories /etc/apk/repositories.disabled
RUN echo 'https://$ARTIFACTORY_USER_PROFILE:$CG_PRIVATE_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-private/' >> /etc/apk/repositories
RUN apk add sed
USER nonroot
EOF
```

> This example will install the `sed` package — from the `cg-private` remote repository — on the image as it is built.

Reviewer: ```suggestion
This example installs the `sed` package from the `cg-private` remote repository on the image.
```

> After creating the Dockerfile, build the image. Note that this command includes the `-f` option to specify that it should build the image with the `Dockerfile.build` file you just created:

Reviewer: After removing the above section .. simplify to just Dockerfile as usual

```shell
docker build -t ar-build -f Dockerfile.build .
```

> Once again, you can confirm that the pull-through and caching worked as expected by checking the **Artifacts** page in the Artifactory dashboard.

Reviewer: ```suggestion
Confirm that the pull-through and caching worked as expected by checking the **Artifacts** page in the Artifactory dashboard.
```

## Configuring Pull-Through Caches for Chainguard's Public Repositories

> Customers also have access to Chainguard's public `chainguard` and `extra-packages` repositories. To set up a pull-through cache for these package repositories on Artifactory, you can follow the same procedure outlined previously for the private APK repository.

Reviewer: ```suggestion
You also have access to the public `chainguard` and `extra-packages` repositories. To set up a pull-through cache for these package repositories on Artifactory, you can follow the same procedure outlined previously for the private APK repository.
```

but without username and password right?

> You will need to create two remote repositories within Artifactory — one for each public package repo — by following the same steps as before: 

Reviewer: ```suggestion
You must create two remote repositories within Artifactory — one for each public package repo — by following the same steps as before: 
```

Then, copy the resulting tokens to create a pair of environment variables. For the `cg-chainguard` repository's token:

```shell
export CG_TOKEN=<cg-chainguard-token>
```

Then for the `cg-extras` repo's token:

```shell
export EXTRAS_TOKEN=<cg-extras-token>
```

Following that, you can use these environment variables, along with a few created earlier in the guide, to create a Dockerfile as you previously did for the `cg-private` repository:

```shell
cat > Dockerfile.repos <<EOF
FROM cgr.dev/chainguard/python:latest-dev
USER root
RUN mv /etc/apk/repositories /etc/apk/repositories.disabled
RUN echo 'https://$ARTIFACTORY_USER_PROFILE:$CG_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-chainguard/' >> /etc/apk/repositories
RUN echo 'https://$ARTIFACTORY_USER_PROFILE:$EXTRAS_TOKEN@$ARTIFACTORY_HOSTNAME.jfrog.io/artifactory/cg-extras/' >> /etc/apk/repositories
RUN apk add c-ares
EOF
```

Following that, build the image:

```shell
docker build -t ar-repos -f Dockerfile.repos .
```

As this command runs, it will install the `c-ares` package into the image.

## Using a Virtual Repository as a Pull-Through Cache

> Artifactory allows you to create what it refers to as *virtual repositories*. A virtual repository is a collection of one or more repositories (such as local, remote, or other virtual repositories) that have the same package type. The benefit of this is that you can access resources from multiple locations using a single logical URL.

Reviewer: ```suggestion
Artifactory allows you to create what it refers to as *virtual repositories*. A virtual repository is a collection of one or more repositories, such as local, remote, or other virtual repositories, that have the same package type. The benefit of this is that you can access resources from multiple locations using a single logical URL.
```

> **Note**: This section provides a high-level overview of how to set up a virtual repository on Artifactory that you can use as a pull-through cache for Chainguard packages. For more information on Artifactory's virtual repositories, please refer to the [official documentation](https://jfrog.com/help/r/jfrog-artifactory-documentation/virtual-repositories).

To get started, create a new virtual repository. From the **Repositories** tab, click the **Create a Repository** button. This time, select the **Virtual** option, but instead of selecting the Alpine package type, select **Generic**.

> You can use the Alpine package type, but [virtual Alpine repositories](https://jfrog.com/help/r/jfrog-artifactory-documentation/virtual-alpine-repositories) will only allow you to pull packages that have already been cached by an associated remote repository. A **Generic** artifact repository will provide more flexibility.

Reviewer: Really .. that sounds like a bug .... I am not reading that from their docs.. how did you get to that info?

Author: That was already in the doc, so wherever it came form I got this detail a long time ago. With some testing just now, though, it seems to be true so I'm inclined to leave it in.

On the **New Virtual Repository** page, enter a key of your choosing into the **Repository Key** field.

Next, you must select existing repositories to include within this virtual repository. If you've been following along with this guide, you can test this out with the `cg-private, `cg-chainguard` and `cg-extras` repositories created previously.

Select your repositories by clicking their respective checkboxes. Then be sure to click the right-pointing chevron to move them to the **Selected Repositories** column. Finally, click the **Create Virtual Repository** button. 

> As before, a window will appear letting you know that the repository was created successfully. Click the **Set Up Generic Client** button at the bottom of this window to retrieve the token you'll need to test whether you can pull images through this repository. 

Reviewer: ```suggestion
Click the **Set Up Generic Client** button at the bottom of the success message window to retrieve the token you need to test whether you can pull packages through this repository. 
```

You'll use this token to create one more Dockerfile in the next section.

> ### Testing pull-through with a virtual package repository

Reviewer: Honestly .. i would even collapse this more now that I see .. create all three repos and the virtual repo in one section

Then have one section that shows a docker build that accesses the group in a build

Author: I'm going to hold off on this for now. Partly because I like the current structure and partly because I want to get this off my plate. But I do think what you suggest might be the ideal.
